src/command_executor.cc

```diff
@@ -75,30 +87,40 @@ void CommandExecutor::WaitChildExit(pid_t pid) {
 }
 
 int CommandExecutor::Execute(const std::string &line) {
-  CommandParseResult command_parse_result =
+  std::vector<CommandParseResult> command_parse_result_list =
       parser_.ParseUserInputLine(line);
-  if (build_in_.Exist(command_parse_result.command)) {
-    return build_in_.Execute(command_parse_result.command,
-                             command_parse_result.args);
-  }
+  bool is_pipe_redirect = command_parse_result_list.size() != 1;
+  int pipe_fds[2];
+  pipe(pipe_fds);
+  for (int i = 0; i < command_parse_result_list.size(); i++) {
+    CommandParseResult command_parse_result = command_parse_result_list[i];
+    if (build_in_.Exist(command_parse_result.command)) {
+      return build_in_.Execute(command_parse_result.command,
```

Comment:
Looks you do not handle build_in correctly, maybe you can try something like `alias | grep ls`. (`alias` will list all alias available, and result `grep ls` should only the lines that contains string `ls`)